### ISSUE TITLE
Supporting locale-specific carbon footprint calculations

### ISSUE THREAD

**Issue by shankari** [e-mission/e-mission-docs#399]
@tms-martins wants to extend the carbon footprint calculation to take into account carbon intensity across multiple regions. The challenge with this is that the calculation currently happens on the phone, and works off summary data. So you don't know how much of the travel happened in which region as part of the calculation.

I believe that the correct long-term solution is to support grouping results by geography in addition to time. The summary generation code already supports grouping by time (e.g. DAILY, WEEKLY, MONTHLY, etc) in `emission/analysis/result/metrics/time_grouping.py`

adding a similar `geo_grouping.py` would allow the results be grouped by regions at various granularities (e.g. city/region/state/country etc). Then, the metrics retrieval code on the phone can ask for the data to be grouped by both time and geo, and apply the intensity map from the correct region before displaying the result.

**Comment by shankari** [comment]
@tms-martins, please let me know if you need more information.

**Comment by shankari** [comment]
> regions at various granularities (e.g. city/region/state/country etc

We already reverse geocode using nominatim, currently just for display (`display_name`) 
https://github.com/e-mission/e-mission-server/blob/master/emission/analysis/intake/cleaning/clean_and_resample.py#L246

should be pretty easy to fill in other information, ideally with an OSM id and not just a name for better matching.

**Comment by tms-martins** [comment]
Thank you @shankari, that is already very helpful. As first step I'm currently working on a branch in my fork, to have the carbon footprints calculated based on vehicle categories from GIS data (and not just an aggregate "IN_VEHICLE");

**Comment by tms-martins** [comment]
> > 
> 
> We already reverse geocode using nominatim

@shankari since nominatim has "an absolute maximum of 1 request per second", how would I address this, short of running my own server? Is there an e-mission server I can use for testing?

**Comment by shankari** [comment]
There is an e-mission copy of nominatim - I will send you the URL via email. It hasn't been updated with fresh data for a few years, but it should work for large scale georegions.

**Comment by tms-martins** [comment]
@shankari:
Taking advantage of the reverse geocode call in [`get_filtered_place`](https://github.com/shankari/e-mission-server/blob/dc60ccdfab34e0be4fe76154911243e648085672/emission/analysis/intake/cleaning/clean_and_resample.py#L253), I am retrieving the country code and and storing it in `filtered_place_data` which is a `Cleanedplace` object. The idea would be to later copy this value to the Metadata object of a place. 
Is this a valid approach, and what would be the best context to copy the country code to the Metadata?

**Comment by shankari** [comment]
> There is an e-mission copy of nominatim - I will send you the URL via email. It hasn't been updated with fresh data for a few years, but it should work for large scale georegions.

The e-mission copy seems to have limited data loaded - unfortunately, I don't remember the details since I had a student set it up and it was a couple of hours ago.

There are other implementations of nominatim that have higher rate limits - e.g. mapquest
https://wiki.openstreetmap.org/wiki/Nominatim#Alternatives_.2F_Third-party_providers

**Comment by shankari** [comment]
@tms-martins that sounds reasonable.

And you don't even need to copy it to the metadata. You can leave it in the data. The parameters for the time query are in the data too. The metadata is for information like the create time, the OS etc.

**Comment by tms-martins** [comment]
> You can leave it in the data.

From what I can see, the data isn't stored in the database, whereas the metadata (or some parts of it) are. Maybe I'm wrong? :) Otherwise, I'd have to ensure that the country code is also stored in the database, at least for each place and/or each trip (start/end), correct?

**Comment by shankari** [comment]
>  Could you point me to where the "data" is written/exported to the database?

The `data` object is wrapped using the wrapper objects in `emission/core/wrapper`
The wrapper for the place object is `emission.core.wrapper.cleanedplace`, which we fill in here
https://github.com/e-mission/e-mission-server/blob/master/emission/analysis/intake/cleaning/clean_and_resample.py#L235

**Comment by shankari** [comment]
> So I'd have to ensure that the country code is also stored within, at least for each place and/or each trip (start/end), correct?

Yup. the obvious choice would be to store it in the place object since that is what it is designed for, but then the metric generation may get more complex.

Another option is to use geopandas. It wasn't yet popular when I was writing this code, so I haven't used it, but I have been looking into it recently. And there appears to be some kind of support for aggregation.
http://geopandas.org/aggregation_with_dissolve.html

They haven't checked in the raw dataset that they use, so I am not sure if this is a benefit over regular pandas, but just thought I'd point it out in case it is :)

**Comment by tms-martins** [comment]
> The wrapper for the place object is `emission.core.wrapper.cleanedplace`, which we fill in here

So I'm still missing something. 

I had added a `country_code` to `Cleanedplace` and populated it in `clean_and_resample` `get_filtered_place` (see [here](https://github.com/tms-martins/e-mission-server/blob/9d60f747ca400ea83bfdb714851e009763aed6a9/emission/core/wrapper/cleanedplace.py#L16) and [here](https://github.com/tms-martins/e-mission-server/blob/9d60f747ca400ea83bfdb714851e009763aed6a9/emission/analysis/intake/cleaning/clean_and_resample.py#L254)).

I reset the pipeline, and run it again. I open a mongoDB terminal and list the collections `Stage_timeseries` and `Stage_analysis_timeseries`. Neither listing has a `country_code`, nor a `readable_name` for that matter (which is defined in the same place). 

So I am wondering where and if these ever get written to the database, if I'm looking at the wrong mongo collection. or making a wrong assumption somewhere.

**Comment by shankari** [comment]
> When looking at the Stage_analysis_timeseries collection I can see both country_code and readable_name in a few entries. So only start and end places have these data, right?

Yup. If you use the e-mission timeseries interface instead of accessing the database directly, this should be objects with key = `"analysis/cleaned_place"`

**Comment by tms-martins** [comment]
@shankari Unfortunately I won't be able to implement this feature in the time I have left, although I believe I understand basically what needs to be done. However, I have written a [feature which allows the user to select the locale-based carbon-footprint dataset](https://github.com/tms-martins/e-mission-phone/tree/select-carbon-dataset-locale) via an option in the settings (which isn't the perfect solution, but could be already quite helpful).

**Comment by shankari** [comment]
@tms-martins, sure please submit the locale-based carbon-footprint selection code as a PR and I will review + merge. Could you also submit the code to store the country code into the place object? Then somebody else who has time can try to merge the two components.

**Comment by tms-martins** [comment]
Thank you @shankari. The PR is [here](https://github.com/e-mission/e-mission-phone/pull/597) and I already identified some minor issues which I will address within a day or two.

> Could you also submit the code to store the country code into the place object?

I will tidy up what I did, publish it and let you know :) So, no actual grouping by country code or API for the client, but at least retrieval and storage of country code in the DB was working

**Comment by tms-martins** [comment]
@shankari My contribution of storing the country code in the `place` object can be found here:
https://github.com/tms-martins/e-mission-server/tree/country_code_data
This uses the existing reverse-geocoding call, and seemingly only stores the country code for start/end places in a trip.
